src/csvlook.c

```diff
@@ -0,0 +1,262 @@
+    }
+
+    LOG_V("%s\n", "Starting column width detection");
+    current_cell = _cells;
+    Cell* cell_end = _cells + cells_found;
+    int column_idx = 0;
+    while (current_cell < cell_end) {
+        if (current_cell->start == NULL) {
+            column_idx = 0;
+            LOG_V("%s\n","New Row");
+        }
+        else {
+            LOG_V("Processing column %d: Prev max %d Cur len %d\n", column_idx, config.column_sizes[column_idx], (int)current_cell->length);
+            /* max(prevmax, curlength) */
+            config.column_sizes[column_idx] = (config.column_sizes[column_idx] < (int)current_cell->length) ? (int)current_cell->length : config.column_sizes[column_idx];
+            /* min(curmax, configmax) */
+            config.column_sizes[column_idx] = (config.column_sizes[column_idx] > preconfig.max_width) ? preconfig.max_width : config.column_sizes[column_idx];
+            column_idx++;
+        }
+        current_cell++;
+    }
+    column_idx = 0;
+    LOG_D("%s\n", "Trying to parse specified widths");
+    char *token;
+    char *widths = preconfig.specified_widths;
+    while ((token = strsep(&widths, ",")) != NULL) {
+        if (token[0] != '\0' && strlen(token) > 0) {
+            int len = atoi(token);
+            if (len > 0) {
+                LOG_D("Overriding column %d to %d\n", column_idx, len);
+                config.column_sizes[column_idx] = len;
+            }
+        }
+
+        column_idx++;
+    }
+
+    LOG_D("%s\n","Done finishing config calcs");
+    for (int i = 0; i < config.column_count; i++) {
+        LOG_D("Column %d width: %d\n", i, config.column_sizes[i]);
+    }
+}
+
+static void output_cells(size_t cells_found, bool last_full) {
+    Cell* current_cell = _cells;
+    Cell* cell_end = _cells + cells_found;
+    int column_idx = 0;
+
+    while (current_cell < cell_end) {
+        if (current_cell->start == NULL) {
+            fwrite(config.newline, sizeof(char), (int)config.newline_length, stdout);
+            column_idx = 0;
+        }
+        else {
+            if (current_cell->length == 0) {
+                fprintf(stdout, "%-*s ", column_idx, "");
+            }
+            else {
+                int printlen = (int)current_cell->length;
+                int elipse = 0;
+                if (printlen > config.column_sizes[column_idx]) {
+                  printlen = config.column_sizes[column_idx]-2;
+                  elipse = 1;
+                }
+                fprintf(stdout, "%.*s", printlen, current_cell->start);
+                if (elipse) {
+                  fprintf(stdout, "..");
+                } else
+                if ((int)current_cell->length < config.column_sizes[column_idx]) {
+                    fprintf(stdout, "%*s", config.column_sizes[column_idx] - (int)current_cell->length, "");
+                }
+                fprintf(stdout, " ");
+            }
+            column_idx++;
+        }
+        current_cell++;
+    }
+    if (last_full) {
```

Comment:
This is an essential choice for performance. There is only one buffer we read into, and write from, requiring callbacks for the overflow case would complicate things. 

I might think about it, since it tends to be an edge case anyway,  but the interface gets a lot more complex, especially how to handle the next iteration after a overflow happened.